listeners/measured.go

```diff
@@ -3,38 +3,88 @@ package listeners
 import (
 	"net"
 	"net/http"
+	"sync"
 	"time"
 
 	"github.com/getlantern/measured"
 )
 
-// Wrapped stateAwareMeasuredListener that genrates the wrapped wrapMeasuredConn
+const (
+	rateInterval = 1 * time.Second
+)
+
+// MeasuredReportFN is a function that gets called to report stats from the
+// measured connection. deltaStats is like stats except that SentTotal and
+// RecvTotal are deltas relative to the prior reported stats. final indicates
+// whether this is the last call for a connection (i.e. connection has been
+// closed).
+type MeasuredReportFN func(ctx map[string]interface{}, stats *measured.Stats, deltaStats *measured.Stats,
+	final bool)
+
+// Wrapped stateAwareMeasuredListener that generates the wrapped wrapMeasuredConn
 type stateAwareMeasuredListener struct {
-	measured.MeasuredListener
+	net.Listener
+	reportInterval time.Duration
+	report         MeasuredReportFN
 }
 
-func NewMeasuredListener(l net.Listener, reportInterval time.Duration, m *measured.Measured) net.Listener {
-	return &stateAwareMeasuredListener{
-		MeasuredListener: *m.Listener(l, reportInterval),
-	}
+func NewMeasuredListener(l net.Listener, reportInterval time.Duration, report MeasuredReportFN) net.Listener {
+	return &stateAwareMeasuredListener{l, reportInterval, report}
 }
 
 func (l *stateAwareMeasuredListener) Accept() (c net.Conn, err error) {
-	c, err = l.MeasuredListener.Accept()
+	c, err = l.Listener.Accept()
 	if err != nil {
 		return nil, err
 	}
-	sac, _ := c.(*measured.Conn).Conn.(WrapConnEmbeddable)
-	return &wrapMeasuredConn{
-		WrapConnEmbeddable: sac,
-		Conn:               c.(*measured.Conn),
-	}, err
+	fs := make(chan *measured.Stats)
+	wc := &wrapMeasuredConn{
+		Conn: measured.Wrap(c, rateInterval, func(mc measured.Conn) {
+			fs <- mc.Stats()
+		}),
+		ctx:        make(map[string]interface{}),
+		finalStats: make(chan *measured.Stats),
+	}
+	sac, _ := c.(WrapConnEmbeddable)
```

Comment:
Ahh makes sense @oxtoacart! The only thing is I don't think `sac` will be `nil` if it's not a `WrapConnEmbeddable` will it?